Port a few portals to libdex

swick opened the pull request:
Just an experiment for now.

There are currently two main pain points:

* Requests/Sessions use threads, so we still need take a lock when dealing with them, but with even more pitfalls because keeping the lock while suspended is not great at all
* The `g-authorize-method` mechanism is currently blocking the main thread when ran in a fiber. This should use async, but also needs locking across suspension points.

The request/session thing could be solved by making it async, because they are local to a portal which will always either use threads or use fibers, but never both.

The async lock thing is a bit more trick /cc @chergert.

swick committed:
permission: Add lookup and store variants which return futures

When a dbus method call is dispatched in a fiber, those functions allow
us to query or store permissions without blocking and without callback
hell.

swick committed:
xdp: Port a few portals to use libdex

swick committed:
request-future: Add a new Request implementation which uses futures

This also allows us to change the design a bit: requests get created by
the dbus method handler functions and are owned. When the ref-count
drops to zero, the Request gets closed and unexported. This allows using
an autoptr in the handler for ownership, without ever having to owne it
somewhere else.

swick committed:
session-future: Add a new Session implementation which uses futures

Similar to the Request, this session implementation uses futures and
doesn't need any external ownership. Unlike a Request, a Session is
supposed to be used in multiple calls, so the portal needs to store it
somewhere to look it up from different method handlers. The new
XdpSessionFutureStore can be used for this purpose.

AdrianVovk commented on src/meson.build:
```diff
@@ -1,39 +1,45 @@
   interface_prefix: 'org.freedesktop.portal',
   namespace: 'XdpDbus',
   docbook: 'portal',
+  extra_args: [f'--extension-path=@libdex_codegen_ext@'],
```
Looks like we can factory out the fstring? Or even the array entirely?

swick committed:
session: Take a XdpAppInfo

It already provides the sender and app id. It is also useful because
this way we don't need a lookup from sender to XdpAppInfo anymore.

swick committed:
app-info: Add a destroyed signal

The XdpAppInfo represents a single sender and (ignoring temporary
references) gets destroyed when the sender drops off the bus. That's a
more targeted signal than subscribing to peer_died, and also again means
we don't have to have a lookup from the sender name to the XdpAppInfo.

swick commented:
@AdrianVovk this one is ready for some high level reviewing

swick committed:
context: Make dispatching DBus interface skeletons in threads explicit

We're going to add dispatching in fibers as well, and we want to replace
the threads with it, so a default to threads is not desired anymore.

This also changes the XDP_CONTEXT_EXPORT_FLAGS_HOST_PORTAL flag which so
far meant "run on the main thread and don't authenticate". Now that the
default flags mean "run on the main thread" we can drop the host portal
flag and instead add XDP_CONTEXT_EXPORT_FLAGS_SKIP_AUTH to skip
authentication.

swick committed:
context: Make it possible to dispatch DBus interface skeletons in fibers

Fibers allow us to handle method calls in a single function which can
yield by awaiting futures. This means we can get rid of threading
without blocking the main thread.

swick committed:
app-info-registry: Make xdp_app_info_registry_lookup_sender private

The notification portal is changed to use the XdpAppInfo destroyed
signal which gives direct access to the XdpAppInfo and means it doesn't
need the lookup from sender.

The usb portal uses the XdpAppInfo from the session object.

swick committed:
app-info-registry: Use per-sender libdex channels to enforce order

By creating a queue and a worker fiber which processes enqueued items in
order, we get full order which extends to worker threads.

The channel gets destroyed as soon as there is no more pending work and
re-created on demand.

The insert operation takes a DexFuture which resolves to an XdpAppInfo,
which allows to "register" the insert operation and complete it later
while everything stays in order.

swick committed:
dex: Add dex_scheduler_spawn_closure

Same as dex_scheduler_spawn() but trampolines into a fiber without
having to create special structures on the way there.

We use the dex namespace and hope that we can upstream this, so lets
guard it properly from the buildsystem, so when it becomes available,
everything still works.

swick committed:
ci: Build libdex from commit from main for tests to pass

libdex has a code path which works without io-uring, but that one
creates a thread in the shared object constructor which then triggers:

    g_log_writer_default_set_use_stderr: assertion 'g_thread_n_created () == 0' failed

It also uses g_debug() in the shared object constructor which means it
will get printed (if G_MESSAGES_DEBUG is set appropriately) before the
glib test framework has a chance to redirect it (to add the `# `
prefix). This breaks the test TAP output.

In a later version of libdex, this shared object constructor will get
dropped and both problems will go away.

Ideally there would be a libdex release with the fixes included so we
could just depend on it from meson, but that's not the case. So for now
this just builds from main without bumping the build requirements
because it only affects test and not production.

swick committed:
permission: Get a deep copy from GVariant directly

Instead of getting a shallow copy and then creating a deep copy from it,
we can just get a deep copy via `^as` from the GVariant, which allows us
to steal the value directly.

swick committed:
context: Unexport portals and drain the main loop on dispose

This is done so that fibers get a chance to handle dex_await returning a
cancelled error and shutting things down appropriately.

swick committed:
tests: Add a test checking shutdown with in-flight requests

swick committed:
tests: Make it possible for tests to capture stderr/stdout of xdp

By default this is turned off, so that it lands in the tests logs, but
in some situations we actually want to know about messages being
produced.
